File: src/main/scala/firelib/robot/ProvidersFactory.cs
using System;
using Fire.Common.Threading;
using QuantLib.Common;

namespace Fire.Common.Robot
{
    public class ProvidersFactory
    {
        public void Create(ModelRuntimeConfig cfg, IThreadExecutor executor,  out ITradeGate tradeGate, out IMarketDataProvider marketDataProvider)
        {
            if (cfg.GatewayType == "IB")
            {
                tradeGate = ReflectionCreator<ITradeGate>.Create("IbTradeGate");
                marketDataProvider = (IMarketDataProvider) tradeGate;
                tradeGate.Configure(cfg.GatewayConfig, cfg.IbContractMapping, executor);
                tradeGate.Start();
                return;
            }
            throw new Exception("not supported gateway type " + cfg.GatewayType);
        }
    }
}

File: execution/src/main/scala/firelib/execution/providersFactory.scala
package firelib.execution

import firelib.common.threading.ThreadExecutor
import firelib.common.tradegate.TradeGate
import firelib.execution.config.ModelExecutionConfig

object providersFactory {
    def create(cfg: ModelExecutionConfig, executor: ThreadExecutor): (TradeGate, MarketDataProvider) = {
        val gate = Class.forName(cfg.tradeGateConfig.gateClassName).newInstance().asInstanceOf[TradeGate with Configurable]
        val marketDataProvider =  gate.asInstanceOf[MarketDataProvider]
        gate.start(cfg.tradeGateConfig.gateParams, executor)
        return (gate, marketDataProvider)
    }
}
